openedx_events/event_bus/__init__.py

```diff
@@ -1 +1,83 @@
 """Classes and utility functions for the event bus."""
+
+from abc import ABC, abstractmethod
+from typing import NoReturn, Optional
+
+from django.conf import settings
+from django.utils.module_loading import import_string
+
+
+class EventBus(ABC):
+    """
+    Parent class for event bus implementations.
+    """
+
+    @abstractmethod
+    def send(
+            *, signal: OpenEdxPublicSignal, topic: str, event_key_field: str, event_data: dict,
+            sync: bool = False,
+    ) -> None:
+        """
+        Send a signal event to the event bus under the specified topic.
+
+        Arguments:
+            signal: The original OpenEdxPublicSignal the event was sent to
+            topic: The event bus topic for the event
+            event_key_field: Path to the event data field to use as the event key (period-delimited
+              string naming the dictionary keys to descend)
+            event_data: The event data (kwargs) sent to the signal
+            sync: Whether to wait indefinitely for event to be received by the message bus (probably
+              only want to use this for testing)
+        """
+        pass
+
+    @abstractmethod
+    def consume_indefinitely(*, topic: str, group_id: str) -> NoReturn:
+        """
+        Consume events from a topic in an infinite loop.
+
+        Events will be converted into calls to Django signals.
+
+        Arguments:
+            topic: The event bus topic to consume from.
+            group_id: The consumer group to participate in.
+        """
+        pass
+
+
+class NoEventBus(EventBus):
+    """
+    Stub implementation to "load" when no implementation is properly configured.
+    """
+
+    def send(
+            *, signal: OpenEdxPublicSignal, topic: str, event_key_field: str, event_data: dict,
+            sync: bool = False,
+    ) -> None:
+        pass
+
+    def consume_indefinitely(*, topic: str, group_id: str) -> NoReturn:
+        pass
+
+
+def _load_event_bus() -> EventBus:
+    """
+    Load the currently configured event bus implementation.
+
+    If configuration is invalid, will return a NoEventBus instance that does nothing.
+    """
+    try:
+        eb = settings.EVENT_BUS
+        loader = import_string(eb['loader'])
+        impl = loader(eb)
+        if isinstance(impl, EventBus):
+            return impl
+        else:
+            warnings.warn(f"Loader {loader} a {type(impl)}, not an EventBus", UserWarning)
+            return NoEventBus()
+    except BaseException as e:
+        warnings.warn(f"Failed to load event bus implementation: {e!r}", UserWarning)
+        return NoEventBus()
+
+# The configured event bus implementation, or a stub instance that does nothing.
+EVENT_BUS = _load_event_bus()
```

Comment:
We had punted on caching settings to ensure we supported remote config changes. If we feel we can do this quickly, you might want to look for patterns in DRF settings, where they use the settings_changed signal from Django to reload: https://github.com/encode/django-rest-framework/blob/master/rest_framework/settings.py.

Reply:
That exactly the sort of thing I'd want, except it looks like it's only called during tests. Do we arrange for it to be signaled by remote-config as well?

Reply:
In event-bus-kafka's implementing class I'd love to be able to do something like this:

```
def __init__(...):
    self.producer = create_producer()
    ...and same for consumer
    setting_changed.connect(lambda: self.reload_config())

def reload_config(self):
    old_producer = self.producer
    self.producer = create_producer()
    threading.Thread(target=lambda: old_producer.flush()).start()
    ...and same for consumer
```
...but that assumes we can actually get such a signal.

Reply:
This is the code I was referring to with the signal: https://github.com/encode/django-rest-framework/blob/a1b35bb44b7c9251c8b7bc995aa6598044f1d3ef/rest_framework/settings.py#L256. Are you saying something different?

Reply:
I'm saying that Django does not send that signal except during tests, so we can't rely on it in a running service.

Reply:
48e3918 introduces `get_consumer()` and `get_producer()`, both memoized, rather than a load-time global. (Just to make testing easier! But maybe that's silly.)